keps/sig-node/4112-passdown-resources-to-cri/README.md

```diff
@@ -0,0 +1,1127 @@
+
+First, the resources required by the containers of a pod are not visible at the
+pod sandbox creation time. This can be problematic for example in the case of
+VM-based runtimes where all resources need to be reserved/prepared when the VM
+(i.e. sandbox) is being created.
+
+Second, the kubelet does not provide complete information about the container
+resources specification of native and extended resources (requests and limits)
+to CRI. However, various use cases have been identified where detailed
+knowledge of all the resources can be utilized in container runtimes for more
+optimal resource allocation to improve application performance and reduce
+cross-application interference.
+
+This KEP proposes CRI API extensions for providing complete view of pods
+resources at sandbox creation, and, providing unobfuscated information about
+the resource requests and limits to container runtimes.
+
+## Motivation
+
+When the pod sandbox is created, the kubelet does not provide the CRI runtime
+any information about the resources (such as native resources, host devices,
+mounts, CDI devices etc) that will be required by the application. The CRI
+runtime only becomes aware of the resources piece by piece when containers of
+the pod are created (one-by-one).
+
+This can cause issues with VM-based runtimes
+(e.g. [Kata containers](https://katacontainers.io/) and [Confidential Containers](https://www.cncf.io/projects/confidential-containers/)) that need to prepare the VM before containers are created.
+
+For Kata to handle PCIe devices properly the CRI needs to tell the kata-runtime
+how many PCIe root-ports or PCIe switch-ports the hypervisor needs to create at
+sandbox creation depending on the number of devices allocated by the containers.
+The PCIe root-port is a static configuration and the hypervisor cannot adjust it
+once the sandbox is created. During container creation the PCIe devices are
+hot-plugged to the PCIe root-port or switch-port. If the number of pre-allocated
+pluggable ports is too low, the attachment will fail (container devices >
+pre-allocated hot-pluggable ports).
+
+In the case of Confidential Containers (uses Kata under the hood with additional
+software components for attestation) the CRI needs to consider the cold-plug aka
+direct attachment use-case. At sandbox creation time the hypervisor needs to
+know the exact number of pass-through devices and its properties
+(VFIO IOMMU group, the actual VFIO device - there can be several devices in a
+IOMMU group, attach to PCIe root-port or PCIe switch-port (PCI-Bridge)).
+In a confidential setting a user does not want to reconfigure the VM
+(creates an attack-vector) on every create container request. The hypervisor
+needs a fully static view of resources needed for VM sizing.
+
+Independent of hot or cold-plug the hypervisor needs to know how the PCI(e)
+topology needs to look like at sandbox creation time.
+
+Updating resources of a container means also resizing the VM, hence the
```

Comment:
I think resources are immutable once set

>  # * spec: Forbidden: pod updates may not change fields other than `spec.containers[*].image`,`spec.initContainers[*].image`,`spec.activeDeadlineSeconds`,`spec.tolerations` (only additions to existing tolerations),`spec.terminationGracePeriodSeconds` (allow it to be set to 1 if it was previously negative)

what resources will trigger a VM resizing?

Reply:
Setting resources are not allowed for ephemeral containers, so pod-resizing is the only use case for now (unless we want to consider the container mutation proposal). https://kubernetes.io/docs/concepts/workloads/pods/ephemeral-containers/#what-is-an-ephemeral-container

BTW, pod level resource spec seems like a good fit for hypervisor use cases.

Reply:
> pod level resource spec seems like a good fit for hypervisor use cases.

We don't want to modify the pod spec. Just get visibility on the CRI level. And not just resources (as `ResourceRequirements`) but also other "resources" (in a wider meaning).